src/plugins/discord/fetchDiscord.js

```diff
@@ -0,0 +1,202 @@
+import {DiscordMirror} from "./mirror";
+import {type endCursor, type pageInfo, Fetcher} from "./fetcher";
+
+const DISCORD_SERVER = "https://discordapp.com/api";
+
+type FetchPlanOptions = {|
+  lastMessageId: string,
+  lastMemberId: string,
+  lastReactionId: string,
+|};
+
+/**
+ * Fetches provided url wwith provided options.
+ * Returns either the response.json or rejects with
+ * a retry flag and an optional message.
+ *
+ */
+function tryDiscordFetch(fetch, url, fetchOptions): Promise<any> {
+  return fetch(url, fetchOptions).then((response) => {
+    switch (response.status) {
+      case 403:
+        return Promise.reject({
+          retry: false,
+          message: `403 Forbidden: bad API username or key?\n${response.url}`,
+        });
+      case 404:
+        return Promise.reject({
+          retry: false,
+          message: `404 Not Found on: ${response.url}; maybe bad serverUrl?`,
+        });
+      case 410:
+        return Promise.reject({retry: false, message: `410 Gone`});
+      case 429:
+        return Promise.reject({retry: true, message: null});
+      default:
+        return Promise.resolve(response.json());
+    }
+  });
+}
+
+/**
+ * Retries fetch if retry = true in rejected tryDiscordFetch response.
+ * Throws an error if message provided in rejected tryDiscordFetch response.
+ *
+ */
+function retryDiscordFetch(fetch, url, fetchOptions) {
+  return new Promise((resolve, reject) => {
+    const operation = retry.operation();
+    operation.attempt(() => {
+      tryDiscordFetch(fetch, url, fetchOptions)
+        .then((result) => {
+          resolve(result);
+        })
+        .catch((error) => {
+          if (error.retry && operation.retry(true)) {
+            return;
+          } else if (error.message) {
+            reject(new Error(error.message));
+          } else {
+            reject(error);
+          }
+        });
+    });
+  });
+}
+
+/**
+ * Returns a function that ties the fetch options and
+ * Discord server url to the supplied endpoint.
+ * Clients can provide bot token here to create a Discord Fetch.
+ */
+export function makeDiscordFetch(fetch: typeof fetch, token: BotToken) {
+  return async (endpoint: string) => {
+    const fetchOptions = {
+      method: "GET",
+      headers: {
+        Accept: "application/json",
+        Authorization: `Bot ${token}`,
+      },
+    };
+    const url = new URL(`${DISCORD_SERVER}/${endpoint}`).href;
+    return retryDiscordFetch(fetch, url, fetchOptions);
+  };
+}
+
+/**
+ * Iteratively calls provided `fetch` for as long as the `hasNextPage`
+ * variable in the `pageInfo` response is true.
+ *
+ */
+export async function paginatedFetch(
+  fetch: (endCursor: endCursor) => Promise<pageInfo>,
+  endCursor: endCursor
+): Promise<void> {
+  let after = endCursor;
+  let cont = true;
+
+  while (cont) {
+    const {hasNextPage, endCursor} = await fetch(after);
+    after = endCursor;
+    cont = hasNextPage;
+  }
+}
```

Comment:
I think this function to hide pagination creates confusing / difficult to understand side-effects code.
Why not use an explicit loop?

```js
while (hasNextPage) {
  // ... have side-effects here
}
```

Bonus points if you're able to figure out and use async generators to convert paginated APIs into a stream :smile: (I'll have some examples of this soon©)

```js
class Fetcher {
  // All messages in a channel, using AsyncGenerator as an abstraction over pagination.
  async *messages(
    channel: Model.Snowflake
  ): AsyncGenerator<Model.Message, void, void> {
    /* ... */
  }
}

// When consuming, use the "new" for await ... of
for await (const message of fetcher.messages(channel)) {
  // ... have side-effects here
}
```

Reply:
Can you elaborate on your suggestion? 

The idea of this function was to 1) have a testable pagination helper, and 2) remove the pagination logic from the site of the fetch implementation, so that the side-effects of calling the network and saving the results are emphasized in the fetching logic. 

Writing a unique while loop at the site of every paginated request is hard to test, has the potential to fail silently, and adds cognitive overhead to the code at those sites. Feels like low-hanging fruit to abstract away into a testable format.

Reply:
I agree, because Discord is consistent with how it implements pagination, it's a good idea to implement pagination once in a testable function. Definitely support that. :+1:

What I don't like about the current approach:
- It forces you to use side-effects.
- It assumes we want to use chronological order (feed `after` in a loop).

The assumption part is OK, but should just be explicit through the name for example.
`async function fetchAllPagesChronologically(...)` just spitballing, but something explicit like that.

---

> the side-effects of calling the network

I don't think network calls necessarily are a side-effect. The async nature of it is wrapped with Promises, and we can have a clear input and output type for our functions.

`async fetchGuild(guildId: Snowflake): Promise<Guild>` shouldn't have side-effects.

---

So with that in mind, and now having generics for the pagination. Here's an idea how we could generalize pagination in a single function.

```js

// Given we have these types now.
export type EndCursor = Model.Snowflake;
export type PageInfo = {|
  +hasNextPage: boolean,
  +endCursor: EndCursor,
|};
export type ResultPage<T> = {|
  +pageInfo: PageInfo,
  +results: $ReadOnlyArray<T>,
|};
```

We can now define these.

```js
// A function that accepts `after` and returns a generic ResultPage.
export type FetchChronologicalPage<T> = (
  after: Model.Snowflake
) => Promise<ResultPage<T>>;

// A function to loop these, and return just an array of T.
export async function fetchAllPagesChronologically<T>(
  fetchPage: FetchChronologicalPage<T>
): Promise<$ReadOnlyArray<T>> {
  // ... while (nextPage), etc.
}

// Usage example:
async function getAllMessages(
  channel: Model.Snowflake
): Promise<$ReadOnlyArray<Model.Message>> {
  // This is our concrete FetchChronologicalPage<Model.Message>
  // We're basically partially applying the method so we don't
  // need to provide `channel` as an argument anymore.
  const fetchPage = (after: Model.Snowflake) =>
    this.fetcher.messages(channel, after);

  return fetchAllPagesChronologically(fetchPage);
}

// End user example:
async function processMessages(channel) {
  const messages = await getAllMessages(channel);
  for (const message of messages) {
    // ... ingest the message
  }
}
```

I wouldn't be awfully excited to see this implementation though.
For a modestly large server fetching _all_ messages in a single channel could mean you're spending half an hour and need to keep 100.000s of messages in-memory.

Because it's now wrapped in a single `fetchAllPagesChronologically()` you're also giving up control over the execution. You'll just have to wait till it's done.

Here's where async generators have significant advantages. In terms of memory footprint and control. They basically allow you to solve pagination once, while streaming the data, rather than keeping everything in memory. And giving control to the caller.

```js

// Alternative AsyncGenerator
async function* chronologicalStream<T>(
  fetchPage: FetchChronologicalPage<T>
): AsyncGenerator<T, void, void> {
  // Start at "after: 0".
  let currentPageInfo = {
    hasNextPage: true,
    endCursor: Models.ZeroSnowflake,
  };

  // Keep looping so long as there's a next page.
  while (currentPageInfo.hasNextPage) {
    // Gets the next page (with await).
    const {results, pageInfo} = await fetchPage(currentPageInfo.endCursor);

    // Give control back to the caller by yielding the results.
    yield* ensureAsync(results);

    // Update our pageInfo.
    currentPageInfo = pageInfo;
  }
}

/**
 * Takes any Iterable (sync/async) and turns it to an AsyncGenerator.
 * Note: this is mostly used just to deal with a Flow bug.
 */
export async function* ensureAsync<T>(
  it: Iterable<T> | AsyncIterable<T>
): AsyncGenerator<T, void, void> {
  // Flow doesn't correctly support delegating to a sync iterator in async context.
  yield* (it: any);
}

// Usage example:
function streamMessages(
  channel: Model.Snowflake
): AsyncGenerator<Model.Message, void, void> {
  // Same as before, this is our concrete FetchChronologicalPage<Model.Message>
  const fetchPage = (after: Model.Snowflake) =>
    this.fetcher.messages(channel, after);

  return chronologicalStream(fetchPage);
}

// End user example:
async function processMessages(channel) {
  const messages = streamMessages(channel);
  for await (const message of messages) {
    // ... ingest the message
  }
}
```

Reply:
Note, I'm an `AsyncGenerator` shill. I'm not requiring you use it. If you would like to, I'm happy to help and write some utility functions, like the `ensureAsync` one.

If you prefer using Promises of arrays, that's ok with me as well.